### PR TITLE
fix(cli): support public library methods in modules

### PR DESCRIPTION
**Pull request by vdrg**

### CODE REVIEW

File: packages/cli/src/deploy/resolveConfig.ts [hunk 1]
```diff
@@ -22,7 +24,20 @@ export async function resolveConfig({
...
+  const requirePath = await findUp("package.json", { cwd: process.cwd() });
+  if (!requirePath) throw new Error("Could not find package.json to import relative to.");
+  const require = createRequire(requirePath);
+
+  const moduleOutDirs = config.modules.flatMap((mod) => {
+    if (mod.artifactPath == undefined) {
+      return [];
+    }
+
+    const moduleOutDir = path.join(require.resolve(mod.artifactPath), "../../");
```
Comment (holic) [inline packages/cli/src/deploy/resolveConfig.ts:36]: this `../..` feels a little magical, can we describe it with a comment?

File: packages/cli/src/deploy/resolveConfig.ts [hunk 2]
```diff
@@ -22,7 +24,20 @@ export async function resolveConfig({
   readonly systems: readonly System[];
   readonly libraries: readonly Library[];
 }> {
-  const libraries = findLibraries(forgeOutDir).map((library): Library => {
+  const requirePath = await findUp("package.json", { cwd: process.cwd() });
```
Comment (holic) [inline packages/cli/src/deploy/resolveConfig.ts:27]:
```suggestion
  const requirePath = await findUp("package.json");
```
I believe the `cwd` default is `process.cwd()`

File: packages/cli/src/utils/getContractArtifact.ts [hunk 3]
```diff
@@ -11,18 +11,24 @@ export type GetContractArtifactResult = {
   deployedBytecodeSize: number;
 };
 
+function HexNotStrict(value: string): value is Hex {
```
Comment (holic) [inline packages/cli/src/utils/getContractArtifact.ts:14]:
```suggestion
function isBytecode(value: string): value is Hex {
```

Comment (holic) [reply packages/cli/src/utils/getContractArtifact.ts:14]: there's a better version of this in `world/ts/node/types.ts` if we wanna export that and reuse it (or lift it into common package0

Comment (vdrg) [reply packages/cli/src/utils/getContractArtifact.ts:14]: Right! Saw that but wanted to keep the changes as small as possible,  but happy to include that here if you think that's better

Comment (vdrg) [reply packages/cli/src/utils/getContractArtifact.ts]: for now will keep this so we can refactor and unify using the arktype one in a separate PR

File: packages/cli/src/deploy/resolveConfig.ts [hunk 4]
```diff
@@ -22,18 +24,34 @@ export async function resolveConfig({
...
+  const requirePath = await findUp("package.json");
+  if (!requirePath) throw new Error("Could not find package.json to import relative to.");
+  const require = createRequire(requirePath);
+
+  const moduleOutDirs = config.modules.flatMap((mod) => {
+    if (mod.artifactPath == undefined) {
+      return [];
+    }
+
+    // Navigate up two dirs to get the contract output directory
```
Comment (holic) [inline packages/cli/src/deploy/resolveConfig.ts:36]: I didn't mean to explain what `../..` does but _why_ we're specifically navigating up two dirs

```suggestion
    // forge artifacts are organized by `{outDir}/{sourceFilename}/{contractName}.json`, so we navigate up two directories to get to `outDir`
```

File: .changeset/gentle-carrots-kneel.md [hunk 5]
```diff
@@ -0,0 +1,7 @@
+---
+"@latticexyz/cli": patch
+"@latticexyz/world-module-erc20": patch
+"@latticexyz/world-modules": patch
+---
+
+World Modules (and contracts imported by them) can now use public library methods.
```
Comment (holic) [inline .changeset/gentle-carrots-kneel.md:7]: I would probably separate these two changesets, with the CLI being something like 

> Added support for deploying public libraries used within modules.

Comment (vdrg) [reply .changeset/gentle-carrots-kneel.md]: done!

File: packages/cli/src/deploy/resolveConfig.ts [hunk 6]
```diff
@@ -22,18 +24,34 @@ export async function resolveConfig({
...
-      name: library.name,
-      abi: contractData.abi,
-      prepareDeploy: createPrepareDeploy(contractData.bytecode, contractData.placeholders),
-      deployedBytecodeSize: contractData.deployedBytecodeSize,
-    };
+  const requirePath = await findUp("package.json");
+  if (!requirePath) throw new Error("Could not find package.json to import relative to.");
+  const require = createRequire(requirePath);
+
+  const moduleOutDirs = config.modules.flatMap((mod) => {
```
Comment (holic) [inline packages/cli/src/deploy/resolveConfig.ts:31]: maybe wrap this in a set so if we install multiple ERC20s, we only need to load the libs once?

```suggestion
  const moduleOutDirs = new Set(config.modules.flatMap((mod) => {
```

or ideally, we add support for multiple out dirs in `findLibraries`

**Review by holic — APPROVED** [review]
nice! thanks for the fix!

File: packages/cli/src/deploy/resolveConfig.ts (hunk 6 above)
Comment (vdrg) [reply packages/cli/src/deploy/resolveConfig.ts:31]: ah makes sense, I had actually thought of deduplicating but then forgot. Will try it withing findLibraries 👍

Comment (vdrg) [reply packages/cli/src/deploy/resolveConfig.ts:31]: https://github.com/latticexyz/mud/pull/3321/files

Let me know if this looks good to you 🙏

**Pull request merged**